In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split, GridSearchCV

from sklearn.linear_model import LogisticRegression

from sklearn.metrics import (
    accuracy_score,
    confusion_matrix,
    classification_report
)

In [2]:
df = pd.read_csv("train.csv")

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [3]:
print("Shape of Dataset:", df.shape)

df.info()

Shape of Dataset: (891, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
# Fill missing Age values with median
df["Age"].fillna(df["Age"].median(), inplace=True)

# Fill missing Embarked values with mode
df["Embarked"].fillna(df["Embarked"].mode()[0], inplace=True)

# Drop Cabin column because it has too many missing values
df.drop("Cabin", axis=1, inplace=True)

/tmp/ipykernel_1094/527831011.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["Age"].fillna(df["Age"].median(), inplace=True)
/tmp/ipykernel_1094/527831011.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try 

In [5]:
df.drop(["PassengerId", "Name", "Ticket"], axis=1, inplace=True)

df.head()

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,Embarked
0,0,3,male,22.0,1,0,7.2500,S
1,1,1,female,38.0,1,0,71.2833,C
2,1,3,female,26.0,0,0,7.9250,S
3,1,1,female,35.0,1,0,53.1000,S
4,0,3,male,35.0,0,0,8.0500,S


In [6]:
df = pd.get_dummies(df, columns=["Sex", "Embarked"], drop_first=True)

df.head()

,Survived,Pclass,Age,SibSp,Parch,Fare,Sex_male,Embarked_Q,Embarked_S
0,0,3,22.0,1,0,7.2500,True,False,True
1,1,1,38.0,1,0,71.2833,False,False,False
2,1,3,26.0,0,0,7.9250,False,False,True
3,1,1,35.0,1,0,53.1000,False,False,True
4,0,3,35.0,0,0,8.0500,True,False,True


In [7]:
# Features (Input)
X = df.drop("Survived", axis=1)

# Target (Output)
y = df["Survived"]

print(X.head())
print(y.head())

   Pclass   Age  SibSp  Parch     Fare  Sex_male  Embarked_Q  Embarked_S
0       3  22.0      1      0   7.2500      True       False        True
1       1  38.0      1      0  71.2833     False       False       False
2       3  26.0      0      0   7.9250     False       False        True
3       1  35.0      1      0  53.1000     False       False        True
4       3  35.0      0      0   8.0500      True       False        True
0    0
1    1
2    1
3    1
4    0
Name: Survived, dtype: int64


In [8]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.30,
    random_state=42
)

print("Training Data:", X_train.shape)
print("Testing Data:", X_test.shape)

Training Data: (623, 8)
Testing Data: (268, 8)


In [9]:
model = LogisticRegression(max_iter=1000)

model.fit(X_train, y_train)

print("Model trained successfully!")

Model trained successfully!


In [10]:
y_pred = model.predict(X_test)

print(y_pred[:10])

[0 0 0 1 1 1 1 0 1 1]


In [11]:
accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)

Accuracy: 0.8097014925373134


In [12]:
cm = confusion_matrix(y_test, y_pred)

print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[136  21]
 [ 30  81]]


In [13]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.87      0.84       157
           1       0.79      0.73      0.76       111

    accuracy                           0.81       268
   macro avg       0.81      0.80      0.80       268
weighted avg       0.81      0.81      0.81       268



## Model Evaluation

The Logistic Regression model achieved an accuracy of **80.97%**.

The classification report shows that the model performs slightly better at predicting passengers who did not survive than those who survived.

The model has a Precision of **0.79**, Recall of **0.73**, and F1-Score of **0.76** for the survived class, indicating that some survivors were missed by the model.

In [14]:
param_grid = {
    "C": [0.01, 0.1, 1, 10, 100],
    "solver": ["liblinear", "lbfgs"]
}

In [15]:
grid = GridSearchCV(
    LogisticRegression(max_iter=1000),
    param_grid,
    cv=5,
    scoring="accuracy"
)

grid.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=LogisticRegression(max_iter=1000),
             param_grid={'C': [0.01, 0.1, 1, 10, 100],
                         'solver': ['liblinear', 'lbfgs']},
             scoring='accuracy')

In [16]:
print("Best Parameters:", grid.best_params_)
print("Best Cross Validation Score:", grid.best_score_)

Best Parameters: {'C': 0.1, 'solver': 'lbfgs'}
Best Cross Validation Score: 0.8089032258064517


In [17]:
best_model = grid.best_estimator_

best_model

LogisticRegression(C=0.1, max_iter=1000)

In [18]:
y_pred_tuned = best_model.predict(X_test)

In [19]:
tuned_accuracy = accuracy_score(y_test, y_pred_tuned)

print("Tuned Accuracy:", tuned_accuracy)

Tuned Accuracy: 0.8171641791044776


In [20]:
print(confusion_matrix(y_test, y_pred_tuned))

[[140  17]
 [ 32  79]]


In [21]:
print(classification_report(y_test, y_pred_tuned))

              precision    recall  f1-score   support

           0       0.81      0.89      0.85       157
           1       0.82      0.71      0.76       111

    accuracy                           0.82       268
   macro avg       0.82      0.80      0.81       268
weighted avg       0.82      0.82      0.81       268



In [22]:
comparison = pd.DataFrame({
    "Model": ["Original", "Tuned"],
    "Accuracy": [
        accuracy_score(y_test, y_pred),
        accuracy_score(y_test, y_pred_tuned)
    ]
})

comparison

,Model,Accuracy
0,Original,0.809701
1,Tuned,0.817164


## Before vs After Hyperparameter Tuning

| Model | Accuracy |
|--------|----------|
| Original Logistic Regression | 80.97% |
| Tuned Logistic Regression | (Your Tuned Accuracy) |

GridSearchCV tested different combinations of hyperparameters and selected the best configuration.

In this case, the tuned model performed similarly to the original model. This shows that the original model was already well configured, and hyperparameter tuning confirmed that only minor adjustments were needed.

Accuracy alone can be misleading because it does not show how well a model predicts each class. In an imbalanced dataset, a model may achieve high accuracy by predicting only the majority class while failing to identify the minority class. Metrics like Precision, Recall, and F1-Score provide a more complete evaluation of model performance.

Hyperparameter tuning helped identify the best combination of model parameters automatically. In this project, the tuned model achieved performance similar to the original model, indicating that the initial model was already close to optimal.

## Model Evaluation After Hyperparameter Tuning

After applying GridSearchCV, the Logistic Regression model achieved an accuracy of **81.72%**, which is slightly better than the original model.

The tuned model improved Precision, meaning it made fewer incorrect positive predictions. Although Recall decreased slightly, the overall model performance became more balanced.

This demonstrates how hyperparameter tuning can optimize model performance by selecting the best parameter combination automatically.

# Conclusion

In this project, I evaluated a Logistic Regression model using multiple performance metrics instead of relying only on accuracy.

I used Precision, Recall, and F1-Score to better understand the model's performance and applied GridSearchCV to tune the model's hyperparameters.

The tuned model achieved an accuracy of **81.72%**, improving slightly over the original model. This task demonstrated the importance of proper model evaluation and hyperparameter tuning in Machine Learning.